### PR TITLE
Abstract state storage, extract CircuitMemoryStorage.

### PR DESCRIPTION
**Pull request by phillbaker**
This PR abstracts state storage in preparation for adding a separate redis-backed storage to address https://github.com/danielfm/pybreaker/issues/10.

This PR is currently missing:
 - [x] documentation
 - [ ] separate tests of the new adapter

I wanted to break up the work and get initial feedback on this direction, if this looks directionally correct, I can finish up the remaining work. Thanks!

### CODE REVIEW

File: src/pybreaker.py [hunk 1]
```diff
@@ -214,6 +230,98 @@ def remove_listener(self, listener):
             self._listeners.remove(listener)
 
 
+class CircuitBreakerStorage(object):
```
Comment (danielfm) [inline src/pybreaker.py]: @phillbaker Don't forget the docstrings when you finished implementing these new classes.

File: src/pybreaker.py [hunk 2]
```diff
@@ -214,6 +230,98 @@ def remove_listener(self, listener):
...
+    @property
+    def opened_at(self):
+        return self._opened_at
+
+    @state.setter
+    def opened_at(self, datetime):
+        self._opened_at = datetime
+
+
+class CircuitRedisStorage(object):
```
Comment (danielfm) [inline src/pybreaker.py]: @phillbaker remember to only allow users to use this storage backend if its dependencies are also installed.

File: src/pybreaker.py (hunk 1 above)
Comment (phillbaker) [inline src/pybreaker.py]: Just added!

**Comment by phillbaker** [comment]
@danielfm added the missing documentation - would you prefer adding explicit tests for the MemoryAdapter? The behavior does look well covered by existing tests.

I pulled the code for the redis implementation out of this PR, I'd like to do that as a followup PR if that makes sense?

**Comment by danielfm** [comment]
@phillbaker 

> ...would you prefer adding explicit tests for the MemoryAdapter? The behavior does look well covered by existing tests.

I agree, we don't need more tests if the ones we already have cover the new implementation just as well.

> I pulled the code for the redis implementation out of this PR, I'd like to do that as a followup PR if that makes sense?

Yes, it does. 😄

**Comment by phillbaker** [comment]
Great - anything else I can do to get this PR merged?

**Comment by danielfm** [comment]
Thanks for your contribution, @phillbaker!

**Pull request merged**